cli/util/uploader.go

```diff
@@ -0,0 +1,360 @@
+		prm.SetFilters(filters)
+
+		objectIDs, err := neofs.ObjectSearch(ctx, clientSDK, priv, containerID.String(), prm)
+		if err != nil {
+			return 0, fmt.Errorf("failed to search blocks from %d to %d: %w", start, end, err)
+		}
+		numOIDs := len(objectIDs)
+		if numOIDs == 0 {
+			return 0, nil // No blocks in this range, continue searching.
+		}
+
+		// Return the max block index found in this batch.
+		return start + numOIDs - 1, nil
+	}
+
+	for ; height > 0; height -= searchBatchSize {
+		startIndex := height - searchBatchSize + 1
+		if startIndex < 0 {
+			startIndex = 0
+		}
+
+		maxInBatch, err := searchBatch(startIndex, height)
+		if err != nil {
+			return 0, err
+		}
+
+		if maxInBatch > 0 {
+			return maxInBatch, nil
+		}
+	}
+	return -1, nil
+}
+
+// updateIndexFiles updates the index files in the container.
+func updateIndexFiles(ctx *cli.Context, p pool.Pool, clientSDK *client.Client, containerID cid.ID, account wallet.Account, currentHeight uint, attributeKey string, blockAttributeKey string) error {
+	indexFileSize := ctx.Uint("index-file-size")
+	fmt.Fprintln(ctx.App.Writer, "Updating index files...")
+	prm := client.PrmObjectSearch{}
+	filters := object.NewSearchFilters()
+	filters.AddFilter(attributeKey, fmt.Sprintf("%d", 0), object.MatchNumGE)
+	prm.SetFilters(filters)
+
+	objectIDs, err := neofs.ObjectSearch(ctx.Context, clientSDK, account.PrivateKey(), containerID.String(), prm)
+	if err != nil {
+		return fmt.Errorf("no OIDs found for index files: %w", err)
+	}
+
+	existingIndexCount := uint(len(objectIDs))
+	expectedIndexCount := currentHeight / uint(indexFileSize)
+
+	if existingIndexCount >= expectedIndexCount {
+		fmt.Fprintf(ctx.App.Writer, "Index files are up to date. Existing: %d, Expected: %d\n", existingIndexCount, expectedIndexCount)
+		return nil
+	}
+
+	var buf bytes.Buffer
+	ctxWithCancel, cancel := context.WithCancel(ctx.Context)
+	defer cancel()
+	workerPool := make(chan struct{}, maxParallelSearches)
+	errCh := make(chan error, 1)
+	wg := sync.WaitGroup{}
+
+	for i := existingIndexCount; i < expectedIndexCount; i++ {
+		startIndex := i * indexFileSize
+		endIndex := startIndex + indexFileSize
+		oidInIndexFile := make([]oid.ID, 0, indexFileSize)
+
+		for j := int(startIndex); j < int(endIndex); j = j + searchBatchSize {
+			remaining := int(indexFileSize) - len(oidInIndexFile)
+			end := j + min(searchBatchSize, remaining)
+			prm = client.PrmObjectSearch{}
+			filters := object.NewSearchFilters()
+			filters.AddFilter(blockAttributeKey, fmt.Sprintf("%d", j), object.MatchNumGE)
+			filters.AddFilter(blockAttributeKey, fmt.Sprintf("%d", end), object.MatchNumLE)
+			prm.SetFilters(filters)
+			objectIDs, err = neofs.ObjectSearch(ctx.Context, clientSDK, account.PrivateKey(), containerID.String(), prm)
+			if err != nil {
+				return fmt.Errorf("no OIDs found for index files: %w", err)
+			}
+			oidInIndexFile = append(oidInIndexFile, objectIDs...)
```

Comment:
OK, I think you don't need this list. Allocate buffer of `indexFileSize*oidSize` bytes before this cycle.  Start the set of processing routines; send every OID to the channel; processing routine should read from this channel and Head/parse object header, get index attribute and write OID to the buffer using specified index and increment thread-safe counter. After main cycle finishes, wait until counter is incremented up to `indexFileSize`. Write ready-to-use buffer to NeoFS.

Reply:
If you have questions, we may discuss it.

Reply:
WaitGroup may be replaced by thread-safe counter to reuse this counter for multiple index files and not to start new goroutines for every index file.